discovery_engine_core/web-api/src/bin/ingestion.rs

```diff
@@ -418,8 +418,16 @@ async fn handle_post_documents(
 
     let start = Instant::now();
 
-    let (documents, failed_documents) = body
+    let (valid_ids_docs, mut invalid_ids) = body
         .documents
+        .into_iter()
+        .map(|document| match document.id.is_id_valid() {
+            true => Ok(document),
+            false => Err(document.id),
+        })
```

Comment:
❓ instead of handling this error during the ingestion, we could already handle it during the routing. for example in `document_properties_path()` where the id string is extracted and the id type is created via `DocumentId::new()`, we could add the regex check within the `DocumentId::new()` consturctor as well. then this would also reject the invalid request early on before much work is done server-side. wdyt?

Reply:
I was thinking about this solution, the question is basically whether we want to refuse whole batch of docs because some (maybe all) document IDs are invalid or do we want to process only the ones with the proper ID (if everything else won't fail). I think that in most cases either all IDs will be valid or all IDs will be invalid. Taking that into the account, the approach proposed by you is more ergonomic, but then should we just return `BAD_REQUEST` if we spot single wrong ID and stop the ingestion at this point?

Reply:
IMHO it's okay the reject the whole batch with a single BAD_REQUEST for the ingestion endpoint if it contains invalid document ids. I also don't think keeping two different error sub-variants for "empty" and "bad chars" makes sense.

There is a clear definition about how the use the API and there is only that much we should do for the case of a obvious wrong API usage. In case of a fully public API like for personalization we probably have to anyway reduce the amount of error details we report but that's a bit off topic.

So preferable all ids would:

- implement `FromStr` (for usage in url path)
- have a check in their deserialization logic for validity 

The later is a bit non obvious to do, you need roughly something like:

```rust
impl<'de> Deserialize<'de> for DocumentId {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let inner = String::deserialize(deserializer)?;
        if is_valid_id(&inner) {
            Ok(Self(inner))
        } else {
            // you probably need  <D::Error as serde::de::Error>::invalid... instead for it to compile
            Err(D::Error::invalid_value(Unexpected::Str(&inner), &"Id must match regext [...]+"))
        }
    }
}
```

With this we:

1. make sure any xxxId is always well formed
2. as long as we pass deserialization errors on (in actix we currently do) the user still gets an err with details for the first malformed document id

Reply:
I decided to go with the deserialization approach suggested by @rustonaut